### ISSUE TITLE
Losing async context in AWS Lambda

### ISSUE THREAD

**Issue by pkmnfrk** [honeycombio/beeline-nodejs#374]
I am adding beeline to an existing NodeJS AWS Lambda project. I had it working wonderfully for a while, but at some point something broke and now my traces are losing their context midway through. I've tested it with a lambda as simple as this, and it keeps breaking:

```
const beeline = require("honeycomb-beeline");
beeline(...);
async function entry(event, context) {
    const trace = beeline.startTrace();
    await (async () => {})();
    beeline.finishTrace(trace); // <-- error
}
```

My actual code is a lot more complicated, obviously. Here is a sample log output. I added some extra debugging statement to dump the context every time it finishes a span, and to dump new spans/contexts when they are created:

https://gist.github.com/pkmnfrk/e26d143d0828a966768040dd5a04a347

I'm happy to provide more information if required

**Comment by gmunguia** [comment]
Have you tried using `flush()`? https://github.com/honeycombio/beeline-nodejs/blob/main/docs/API.md#flushing-events

**Comment by vreynolds** [comment]
Hey @pkmnfrk , that's a lot of good info, thanks for being thorough!

Since you mentioned async calls in Slack, I'm curious: how are you starting spans in your code? (Specifically interested if you're using [AsyncSpans](https://docs.honeycomb.io/getting-data-in/javascript/beeline-nodejs/#asynchronous-spans))

I tried to reproduce using your simplified code sample, but coming up short of errors 😞 Is there anything else in your test function runtime, dependency or otherwise?

**Comment by pkmnfrk** [comment]
Hi @vreynolds,

We are using async spans. And, I've double checked that I'm using them correctly (I think). However, I was able to create a minimal test case that reproduces the problem even outside of Lambda, which strongly suggests that I'm not using them correctly.

https://github.com/pkmnfrk/honeycomb-test

Running this code in the terminal, I get the same error. Is there something special we need to do to use traces in an async context?

**Comment by pkmnfrk** [comment]
The full output of running that locally is:

```
mike.caron@PP honeycomb-test % node index    
  honeycomb-beeline:event using impl: libhoney-event +0ms
  honeycomb-beeline:event finishing span {"payload":{"name":"delay","trace.trace_id":"b5ff4b460550054e7b9c510aa4d7daf9","trace.span_id":"1cad1a7dbf87c08f","trace.parent_id":"3a28eda430f3005e"},"startTime":1624294696344,"startTimeHR":[3286968,186360242]} +0ms
  honeycomb-beeline:event enqueuing presampled event ev = {"meta.local_hostname":"PP.LPTP-C02YTHCMLVCG","service_name":"honeycomb-test","name":"delay","trace.trace_id":"b5ff4b460550054e7b9c510aa4d7daf9","trace.span_id":"1cad1a7dbf87c08f","trace.parent_id":"3a28eda430f3005e","duration_ms":1004.240691,"meta.instrumentations":[],"meta.instrumentation_count":0,"meta.beeline_version":"2.7.1","meta.node_version":"v12.22.1"} +1ms
  honeycomb-beeline:event finishing span {"payload":{"name":"entry","trace.trace_id":"b5ff4b460550054e7b9c510aa4d7daf9","trace.span_id":"3a28eda430f3005e"},"startTime":1624294696343,"startTimeHR":[3286968,185923320]} +0ms
  honeycomb-beeline:event -------------------
  honeycomb-beeline:event     honeycomb-beeline error: no payload for span we're trying to finish (stack is empty).
  honeycomb-beeline:event     please paste this message (everything between the "----" lines) into an issue
  honeycomb-beeline:event     at https://github.com/honeycombio/beeline-nodejs/issues.  feel free to edit
  honeycomb-beeline:event     out any application stack frames if you'd rather not share those
  honeycomb-beeline:event     Error
  honeycomb-beeline:event     at LibhoneyEventAPI.askForIssue (/Users/mike.caron/code/honeycomb-test/node_modules/honeycomb-beeline/lib/api/libhoney.js:344:7)
  honeycomb-beeline:event     at LibhoneyEventAPI.finishSpan (/Users/mike.caron/code/honeycomb-test/node_modules/honeycomb-beeline/lib/api/libhoney.js:210:12)
  honeycomb-beeline:event     at LibhoneyEventAPI.finishTrace (/Users/mike.caron/code/honeycomb-test/node_modules/honeycomb-beeline/lib/api/libhoney.js:120:10)
  honeycomb-beeline:event     at Function.finishTrace (/Users/mike.caron/code/honeycomb-test/node_modules/honeycomb-beeline/lib/api/index.js:101:20)
  honeycomb-beeline:event     at Object.entry (/Users/mike.caron/code/honeycomb-test/handler.js:36:17)
  honeycomb-beeline:event     current request context:
  honeycomb-beeline:event     ------------------- +1ms
done { statusCode: 200, body: '{"ok":true}' }
```

**Comment by pkmnfrk** [comment]
One other aspect is that if I switch to non-async spans (which is only practical by using `startSpan`/`finishSpan`, not `withSpan`), then it seems to work correctly.

https://github.com/pkmnfrk/honeycomb-test/tree/nonAsync

Is this expected?

**Comment by pkmnfrk** [comment]
Ok, last comment. I updated the `master` and `nonAsync` branches with a more complicated example. The `master` branch, using `startAsyncSpan`. still shows errors (more than one, now). The `nonAsync` branch seems to be working okay.

I think the takeaway here is that `startAsyncSpan` is more for callback-based async, not Promise-based async. If we can guarantee the order that spans complete, we are better off using non-async spans, right?

**Comment by vreynolds** [comment]
Thanks for the exhaustive example, super helpful! Async spans _should_ work with async/await, but I do see the error in the example you shared, going to dig into that.

In general, you are correct in that as long as you can guarantee the order of spans finishing, you shouldn't need async spans.

**Comment by pkmnfrk** [comment]
Thanks for the confirmation, I was able to take this and start using regular spans everywhere to great effect. Unfortunately, our full codebase is a hodge podge of callback- and promise-based async, so I'm going to have to be very careful. 😅 

Also, good luck finding and fixing this. Based on the research I did thus far, my brain is not big enough to understand how node propagates async context, or how to use that async API. It's unfortunate that you have to support node 10.x, they added an ASP.NET-like context object in 12.x that seems _significantly_ easier to use.

**Comment by vreynolds** [comment]
Oof, @pkmnfrk, this is an adventure! I've poked at this a few times, and so far can only confirm that with certain combos of async witchery `async_hooks.executionAsyncId()` switches between span start and finish. The result is unexpected, from my understanding of async hooks, but the async hooks docs are... sparse 😭

I did have a curious discovery: in your example, the `delay` function in `complicated.js` does not have the `async` modifier, but is being awaited in usages. When I add `async` to `delay` signature, the error goes away, and the execution async IDs are as I expect them. Is that something that could be the issue in your actual code as well?
